auth/scoperesolver.js

```diff
@@ -0,0 +1,338 @@
+    // If a client was loaded add it back
+    if (client) {
+      // For reasoning on structure, see reload()
+      let lastUsedDate = new Date(client.details.lastDateUsed);
+      this._clients.push({
+        clientId:       client.clientId,
+        accessToken:    client.accessToken,
+        updateLastUsed: lastUsedDate < taskcluster.fromNow('-6h')
+      });
+    }
+    this._computeFixedPoint();
+  }
+
+  async reloadRole(roleId) {
+    let role = await this._Role.load({roleId}, true);
+    // Always remove it
+    this._roles = this._roles.filter(r => r.roleId !== roleId);
+    // If a role was loaded add it back
+    if (role) {
+      let scopes = role.scopes;
+      if (!role.roleId.endsWith('*')) {
+        // For reasoning on structure, see reload()
+        scopes = _.union(scopes, ['assume:' + role.roleId]);
+      }
+      this._roles.push({roleId: role.roleId, scopes});
+    }
+    this._computeFixedPoint();
+  }
+
+  async reload() {
+    debug("Loading clients and roles");
+
+    // Load clients and roles in parallel
+    let clients = [];
+    let roles   = [];
+    await Promise.all([
+      // Load all clients on a simplified form:
+      // {clientId, accessToken, updateLastUsed}
+      // _computeFixedPoint() will construct the `_clientCache` object
+      this._Client.scan({}, {
+        handler: client => {
+          let lastUsedDate = new Date(client.details.lastDateUsed);
+          clients.push({
+            clientId:       client.clientId,
+            accessToken:    client.accessToken,
+            // Note that lastUsedDate should be updated, if it's out-dated by more
+            // than 6 hours (cheap way to know if it's been used recently)
+            updateLastUsed: lastUsedDate < taskcluster.fromNow('-6h')
+          });
+        }
+      }),
+      // Load all roles on a simplified form: {roleId, scopes}
+      // _computeFixedPoint() will later add the `expandedScopes` property
+      this._Role.scan({}, {
+        handler(role) {
+          let scopes = role.scopes;
+          if (!role.roleId.endsWith('*')) {
+            // Ensure identity, if role isn't a prefix pattern. Obviously,
+            // 'assume:ab' which matches 'assume:a*' doesn't have 'assume:a*'
+            // by the identity relation. But for non-prefix patterns, the
+            // identify relation implies that you have 'assume:<roleId>'.
+            // This speeds up fixed-point computation, and means that if you
+            // have a match without any *, then you can look up the role, and
+            // not have to worry about any prefix patterns that may also match
+            // as they are already saturated.
+            scopes = _.union(scopes, ['assume:' + role.roleId]);
+          }
+          roles.push({roleId: role.roleId, scopes});
+        }
+      })
+    ]);
+
+
+    // Set _roles and _clients at the same time and immediately call
+    // _computeFixedPoint, so anyone using the cache is using a consistent one.
```

Comment:
In Python there are tools to serialize and debounce async functions like this, so that at least one invocation of the function *starts* after the last invocation.  Presumably similar things in JS?